runtime/error.c

```diff
@@ -1,12 +1,128 @@
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
 #include "balrt.h"
+#include "third-party/libbacktrace/backtrace.h"
+
+#define INITIAL_PC_COUNT 1
+#define MAX_PC_COUNT 256
+#define SKIP_FROM_END 1 // This number counts from the last stack frame
+#define THREAD 0
+#define SUCCESS 0
+#define FAIL 1
+
+struct backtrace_state *state = NULL;
+
+typedef struct {
+    uint32_t nPCs;
+    uint32_t szPCs;
+    bool errorOccurred;
```

Comment:
errorOccurred should separately keep track of whether

- MAX_PC_COUNT was exceeded
- whether an error was returned from libbacktrace; if so, the msg and the errnum (check if msg needs to be copied)

Printing the backtrace should print something different for each case:
- `...` for the former
- the error message from libbacktrace in the latter case

Reply:
In addition to the two kinds of error described above, there's also the possibility of an allocation failure from malloc or realloc.

Reply:
@jclark, I have two unclear points here.
1. When looking at libbacktrace source code, mostly they are using 0 and -1 as errnum. But for each errnum, the error message is not unique. So I think it is not useful to keep libbacktrace errnum in our struct. We can keep only the message. (We have to copy the message because buffer is invalid after function returns)
2. Do we need to keep a list of errors in our struct? Since in backtrace_simple or backtrace_pcinfo functions, there is a possibility to call error_callback several times.

If we are not going to print libbacktrace specific error, we have only three types of error
1. MAC_PC_COUNT exceeds
2. allocation failure
3. libbacktrace internal error

Reply:
The header file has this to say about errnum:

>  The ERRNUM argument, if
   greater than 0, holds an errno value.  The MSG buffer may become
   invalid after this function returns.

 >  As a special case, the ERRNUM argument will be passed as -1 if no
   debug info can be found for the executable, or if the debug info
   exists but has an unsupported version, but the function requires
   debug info (e.g., backtrace_full, backtrace_pcinfo).  The MSG in
   this case will be something along the lines of "no debug info".
   Similarly, ERRNUM will be passed as -1 if there is no symbol table,
   but the function requires a symbol table (e.g., backtrace_syminfo).
   This may be used as a signal that some other approach should be
   tried.

So, errnum is useful: it gives you the errno when libbacktrace failed because of a failed system/library call. But we need msg also (which, as you say, should be copied).

I think a list of errors is not necessary.

Reply:
Changed